README.md
# force-dev-tool

> Command line tool supporting the Force.com development lifecycle

[![Build Status](https://travis-ci.org/amtrack/force-dev-tool.svg?branch=master)](https://travis-ci.org/amtrack/force-dev-tool) OS X / Linux

[![Build Status Windows](https://ci.appveyor.com/api/projects/status/wmdv146qn3oi582u/branch/master?svg=true)](https://ci.appveyor.com/project/amtrack/force-dev-tool/branch/master) Windows


## Install

```console
$ npm install --global force-dev-tool
```

## Usage

```console
$ force-dev-tool --help
force-dev-tool.

Usage:
  force-dev-tool <command> [<args>...]
  force-dev-tool -h | --help
  force-dev-tool --version

Options:
  -h --help       Show this screen.
  --version       Show version.

Commands:
  help            Print help for a command or in general
  remote          Manage orgs (list, add, remove, set default, set password)
  login           Login using Metadata API and show login URL
  fetch           Fetch describe information from a remote
  info            Show describe information from a remote
  package         Generate a package.xml file from local describe information
  retrieve        Retrieve metadata specified in package.xml
  deploy          Deploy metadata specified in a package.xml
  deployTest      DEPRECATED! Use `deploy -t` instead
  validate        DEPRECATED! Use `deploy -c` instead
  validateTest    DEPRECATED! Use `deploy -ct` instead
  test            Execute unit tests
  changeset       Create a changeset/deployment from a unified diff input or cli args
  query           Execute a SOQL query returing JSON
  bulk            (alpha) Import/export data in CSV format using the bulk API
  execute         (alpha) Execute anonymous Apex
  completion      (alpha) Print command line completion

See 'force-dev-tool help <command>' for more information on a specific command.
```

## Examples
**Managing remote environments**

```console
$ force-dev-tool remote add mydev user pass --default
$ force-dev-tool remote add build user pass2
$ force-dev-tool remote add production user pass3 https://login.salesforce.com
```

**Validating credentials for a given remote (optional)**

```console
$ force-dev-tool login mydev
Logged in successfully to remote mydev.
Use the following URL to open Salesforce in your web browser:

https://mynamespace.my.salesforce.com/secur/frontdoor.jsp?sid=REDACTED
```

**Building a manifest**

Fetch various information from the remote first

```console
$ force-dev-tool fetch --progress
Fetching from remote mydev
API Versions
Available Metadata Types
Folders
Metadata Components
RecordTypes of PersonAccount
Active Flow versions
Created config/mydev-fetch-result.json
Fetching remotes finished.
```

Now generate a `package.xml` file based on the fetched information

```console
$ force-dev-tool package -a
Created src/package.xml
```

In order to exclude certain metadata components from being added to the `package.xml` file, add patterns (similar to `.gitignore`) to `.forceignore`. See [here](https://gist.github.com/amtrack/7b99d31b60971b95dda801fd58288257) for some sane default rules.

**Retrieving metadata**

```console
$ force-dev-tool retrieve
Retrieving from remote mydev to directory src
Succeeded
```

**Creating deployments**

1\. By explicitly listing metadata files or metadata components
```console
$ force-dev-tool changeset create vat src/pages/AccountExtensionVAT.page CustomField/Account.VAT__c
```

2\. By providing a unified diff (e.g. `git diff`). Tweak the `git diff` command with `--ignore-space-at-eol` or `--ignore-all-space` to ignore space changes.
```console
$ git diff --no-renames master feature/vat | force-dev-tool changeset create vat
```

Both approaches lead to the following result
```console
Manifest:
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
 <types>
     <members>Account.VAT__c</members>
     <name>CustomField</name>
 </types>
 <types>
     <members>AccountExtensionVAT</members>
     <name>ApexPage</name>
 </types>
 <version>38.0</version>
</Package>

exported metadata container to config/deployments/vat
```

**Creating destructive deployments (reverting changes)**

1\. By explicitly listing metadata files or metadata components
```console
$ force-dev-tool changeset create undo-vat --destructive src/pages/AccountExtensionVAT.page CustomField/Account.VAT__c
```

2\. By providing a unified diff (e.g. `git diff`)
```console
$ git diff --no-renames feature/vat master | force-dev-tool changeset create undo-vat
```

Both approaches lead to the following result
```console
Manifest:
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
 <version>38.0</version>
</Package>

Destructive Manifest:
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
 <types>
     <members>Account.VAT__c</members>
     <name>CustomField</name>
 </types>
 <types>
     <members>AccountExtensionVAT</members>
     <name>ApexPage</name>
 </types>
</Package>

exported metadata container to config/deployments/undo-vat
```

**Deploying metadata**

```console
$ force-dev-tool validate
$ force-dev-tool validateTest
$ force-dev-tool validateTest -d config/deployments/vat
$ force-dev-tool deploy
$ force-dev-tool deployTest
```

Options:

	-d=<directory>    Directory containing the metadata and package.xml [default: ./src].

**Running unit tests**

Running all local tests

```console
$ force-dev-tool test
Running Test execution to remote mydev
Failures:
Test_Foo#test_method_one took 32.0
  - System.AssertException: Assertion Failed: Expected: foo, Actual: bar
  - Class.Test_Foo.test_method_one: line 8, column 1
Test_Foo2#test_method_one took 11.0
  - System.AssertException: Assertion Failed
  - Class.Test_Foo2.test_method_one: line 7, column 1
Error: Visit https://mynamespace.my.salesforce.com/changemgmt/monitorDeploymentsDetails.apexp?asyncId=REDACTED for more information.
3 methods, 2 failures
```

Running specified test classes

```console
$ force-dev-tool test --classNames 'Test_MockFoo Test_MockBar'
```

Running test classes matching a pattern (in src/package.xml)

```console
$ force-dev-tool package grep 'ApexClass/Test_Mock*' \
 | cut -d '/' -f 2 \
 | xargs -0 force-dev-tool test --classNames
```

**Using `force-dev-tool` in a build script**

The following environment variables can be used to define a default remote environment called `env`:

* `SFDC_USERNAME`
* `SFDC_PASSWORD`
* `SFDC_SERVER_URL`

```console
$ force-dev-tool validateTest env
```

Note: You can also define named remotes using [Environment Variables](https://github.com/amtrack/force-dev-tool/wiki/Environment-Variables) (e.g. `SFDC_ci_USERNAME`, `SFDC_ci_PASSWORD`, `SFDC_ci_SERVER_URL`).

**Executing a SOQL query**

```console
$ force-dev-tool query "SELECT Id, Name FROM Account LIMIT 1"
[
  {
    "attributes": {
      "type": "Account",
      "url": "/services/data/v38.0/sobjects/Account/001200000183ZCFAA2"
    },
    "Id": "001200000183ZCFAA2",
    "Name": "GenePoint"
  }
]

$ force-dev-tool query "SELECT COUNT(Id) c FROM Account"
[
  {
    "attributes": {
      "type": "AggregateResult"
    },
    "c": 15
  }
]
```

**(alpha) Exporting/importing data using the bulk API**

Exporting data

```console
$ force-dev-tool bulk export "SELECT Id, Name FROM Account LIMIT 1"
"Id","Name"
"001200000183ZCFAA2","GenePoint"
$ force-dev-tool bulk export "SELECT Id, Name FROM Account" --out Accounts.csv
```

Updating data

```console
$ force-dev-tool bulk update Account --in Accounts.csv --out Accounts-update-results.csv
```

Note: Importing more than one batch is currently not yet supported.

**(alpha) Executing anonymous Apex**

```console
$ echo "insert new Account(Name = 'Test Account');" | force-dev-tool execute
```

## Getting help

Please see the [wiki](https://github.com/amtrack/force-dev-tool/wiki) for [Motivation](https://github.com/amtrack/force-dev-tool/wiki/Motivation) and [Troubleshooting](https://github.com/amtrack/force-dev-tool/wiki/Troubleshooting) and [Resources](https://github.com/amtrack/force-dev-tool/wiki/Resources).

Feel free to open issues with questions.

## License
MIT © [Matthias Rolke](mailto:mr.amtrack@gmail.com)


CHANGELOG.md
# CHANGELOG

## 1.1.1
Fixed
 * changeset: handle recordType in layoutAssignments of Profile @leboff (#126)

## 1.1.0
Added
 * implement multiple packaging options (`-p, --no-recursive` and `-m, --include-managed`)
 * implement `remote display` command to show `sfdx`-like JSON
 * add some unit and integration tests for changeset command
 * deploy/test: show overall code coverage when available

Changed
 * reimplement changeset command using streams @leboff (#99)
 * bump default API version to 41.0
 * print to stderr on failure

Fixed
 * handle DashboardFolder, DocumentFolder, EmailFolder and ReportFolder (#103)
 * TypeError: field.childNamed is not a function (#97)
 * improve package generation for translation types

## 1.0.0
Removed (breaking change)
 * drop support for node 0.12

Added
 * implement querying using Tooling API
 * .forceignore: implement whitelisting
 * add package-lock.json to speed up installation

Fixed
 * fix standard picklists
 * fix regex alerted by eslint
 * adjust timeout of flaky unit test
 * xmldoc: use main repository again and fix indentation for metadata

Changed
 * update dev dependencies
 * clarify usage of environment variables
 * use metadata-writer for writing packages

## 0.13.1

 * fix sort order of childXmlTypes in metadata file container
 * declare EscalationRule as childType

## 0.13.0

 * fix package.xml generation (#82)
 * implement `info list`, `info grep` and `info show` subcommands
 * implement `package add`, `package remove` and `package list` subcommands
 * implement retrieving packages (outbound changesets)
 * add support for more DeployOptions: `--runTests` (#76) and `--purgeOnDelete`
 * deprecate aliases for `deploy`: `deployTest`, `validate` and `validateTest`

## 0.12.0

 * fix running specific tests (#80)
 * implement `package grep` subcommand (#77)
 * implement workaround for standard picklists (#75)

## 0.11.0

 * fix PersonAccount RecordType
 * add info command to README
 * bump API version to 38.0

## 0.10.0

 * add --classNames option to test command (#42)

## 0.9.0

 * fix `npm run fmt` for macOS
 * fix getApiVersion() of fetch-result-parser
 * implement zipping and unzipping based on yazl and yauzl
	* fixes issues deploying metadata with file names containing umlauts
	* adds ability to deploy zip files instead of directories
 * add `package version` command to get/set api version of a `package.xml`
 * add `info` command to show describe information for a remote

## 0.8.0

 * improve generating the package.xml file by fetching more information from remotes
	* determine active Flow version by running a SOQL query
	* identify PersonAccount RecordTypes listed as Account RecordTypes by running a SOQL query
	* retrieve api versions from remote
 * improve creating changesets
	* implement diffing profiles and permissionsets

## 0.7.1

 * fix issue with minor release of vinyl npm package

## 0.7.0

 * implement custom sort order for `package.xml`
 * update default API version to `37.0`
 * add integration test to validate deploying metadata
 * update dependencies

## 0.6.0

 * add `login` command to log in using the Metadata API
 * add example for creating a destructive deployment to README

## 0.5.0
 * drop support for node v0.10
 * format code using js-prettify
 * refactoring
 * improve code coverage
 * add integration tests for cli

## 0.4.0
 * add `query` command for running a SOQL query
 * add `bulk` command for importing/exporting data
 * add `execute` command for executing anonymous Apex
 * remove undocumented `backup` command
 * `force-dev-tool` can now handle being called from within a subdirectory of the project (e.g. `src/`)
 * import force-metadata-utils package for making maintenance easier
 * update dependencies

## 0.3.2
 * allow overwriting deployment directory if `--force` flag is provided
 * improve stability of `remote` command
 * make `fetch` command work with PE editions where Apex is not available

## 0.3.1
 * fix `changeset create` command with binaries
 * include `codeCoverageWarnings` in output of `deploy` command

## 0.3.0
 * add support for windows

## 0.2.0
 * implement running unit tests

## 0.1.2
 * fix deploy command

## 0.1.1
 * explicitly specify test level 'RunLocalTests' for test execution

## 0.1.0
 * intitial release


lib/config.js
"use strict";

var config;

var Config = module.exports = function(opts) {
	opts = opts ? opts : {};
	if (config === undefined) {
		config = {};
		config.pollTimeout = opts.pollTimeout !== undefined ? opts.pollTimeout : 15 * 60 * 1000; // 15 minutes
		config.defaultMetadataTypes = opts.defaultMetadataTypes !== undefined ? opts.defaultMetadataTypes : ['ApexClass', 'ApexComponent', 'ApexPage', 'ApexTrigger', 'CustomField', 'CustomObject', 'CustomTab', 'WebLink'];
		config.defaultApiVersion = opts.defaultApiVersion !== undefined ? opts.defaultApiVersion : "41.0";
		config.folderBasedMetadataMap = opts.folderBasedMetadataMap !== undefined ? opts.folderBasedMetadataMap : {
			'EmailFolder': 'EmailTemplate',
			'DashboardFolder': 'Dashboard',
			'DocumentFolder': 'Document',
			'ReportFolder': 'Report'
		};
		// https://developer.salesforce.com/docs/atlas.en-us.packagingGuide.meta/packagingGuide/packaging_component_attributes.htm
		// Components with no entry in the "Subscriber and Developer Editable" column
		config.nonEditableManagedMetadataTypes = ['ApexClass', 'SharingReason', 'ApexTrigger', 'CompactLayout', 'CustomLabel', 'CustomPermission', /* CustomSetting */ 'FlexiPage', 'HomePageComponent', 'AuraDefinitionBundle', 'PermissionSet', 'PlatformCachePartition', 'StaticResource', 'Translation', 'ApexComponent', 'ApexPage'];
		// https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/standardvalueset_names.htm
		config.standardPicklistMapping = {
			'AccountContactRelation.Roles': 'AccountContactMultiRoles',
			'AccountContactRole.Role': 'AccountContactRole',
			'Account.Ownership': 'AccountOwnership',
			'Account.Rating': 'AccountRating',
			'Lead.Rating': 'AccountRating',
			'Account.Type': 'AccountType',
			'Asset.Status': 'AssetStatus',
			'CampaignMember.Status': 'CampaignMemberStatus',
			'Campaign.Status': 'CampaignStatus',
			'Campaign.Type': 'CampaignType',
			'CaseContactRole.Role': 'CaseContactRole',
			'Case.Origin': 'CaseOrigin',
			'Case.Priority': 'CasePriority',
			'Case.Reason': 'CaseReason',
			'Case.Status': 'CaseStatus',
			'Case.Type': 'CaseType',
			'OpportunityContactRole.Role': 'ContactRole',
			'ContractContactRole.Role': 'ContractContactRole',
			'Contract.Status': 'ContractStatus',
			'Entitlement.Type': 'EntitlementType',
			'Event.Subject': 'EventSubject',
			'Event.Type': 'EventType',
			'Period.PeriodLabel': 'FiscalYearPeriodName',
			'FiscalYearSettings.PeriodPrefix': 'FiscalYearPeriodPrefix',
			'Period.QuarterLabel': 'FiscalYearQuarterName',
			'FiscalYearSettings.QuarterPrefix': 'FiscalYearQuarterPrefix',
			// 'IdeaTheme.Categories': 'IdeaCategory', // You can’t read or update this standard value set or picklist field.
			'Idea.Categories': 'IdeaMultiCategory',
			'Idea.Status': 'IdeaStatus',
			'IdeaTheme.Status': 'IdeaThemeStatus',
			'Account.Industry': 'Industry',
			'Lead.Industry': 'Industry',
			'Account.AccountSource': 'LeadSource',
			'Lead.AccountSource': 'LeadSource',
			'Opportunity.Source': 'LeadSource',
			'Lead.Status': 'LeadStatus',
			'Opportunity.Competitors': 'OpportunityCompetitor',
			'Opportunity.StageName': 'OpportunityStage',
			'Opportunity.Type': 'OpportunityType',
			'Order.Type': 'OrderType',
			'Account.PartnerRole': 'PartnerRole',
			'Product2.Family': 'Product2Family',
			// 'Question.Origin': 'QuestionOrigin', // You can’t read or update this standard value set or picklist field.
			'QuickText.Category': 'QuickTextCategory',
			'QuickText.Channel': 'QuickTextChannel',
			'Quote.Status': 'QuoteStatus',
			'OpportunityTeamMember.TeamMemberRole': 'SalesTeamRole',
			'UserAccountTeamMember.TeamMemberRole': 'SalesTeamRole',
			'UserTeamMember.TeamMemberRole': 'SalesTeamRole',
			'AccountTeamMember.TeamMemberRole': 'SalesTeamRole',
			'Contact.Salutation': 'Salutation',
			'Lead.Name': 'Salutation',
			'ServiceContract.ApprovalStatus': 'ServiceContractApprovalStatus',
			'SocialPost.Classification': 'SocialPostClassification',
			'SocialPost.EngagementLevel': 'SocialPostEngagementLevel',
			'SocialPost.ReviewedStatus': 'SocialPostReviewedStatus',
			'Solution.Status': 'SolutionStatus',
			'Task.Priority': 'TaskPriority',
			'Task.Status': 'TaskStatus',
			'Task.Subject': 'TaskSubject',
			'Task.Type': 'TaskType',
			'WorkOrderLineItem.Status': 'WorkOrderLineItemStatus',
			'WorkOrder.Priority': 'WorkOrderPriority',
			'WorkOrder.Status': 'WorkOrderStatus'
		};
	}
};

Config.prototype.get = function(key) {
	return config[key];
};

Config.prototype.set = function(key, value) {
	config[key] = value;
};
